planner/core/find_best_task.go

```diff
@@ -747,6 +748,13 @@ func (is *PhysicalIndexScan) initSchema(id int, idx *model.IndexInfo, idxExprCol
 func (is *PhysicalIndexScan) addPushedDownSelection(copTask *copTask, p *DataSource, path *accessPath, finalStats *property.StatsInfo) {
 	// Add filter condition to table plan now.
 	indexConds, tableConds := path.indexFilters, path.tableFilters
+
+	tableConds, copTask.rootTaskConds = splitSelCondsWithVirtualColumn(tableConds)
```

Comment:
TiKV can't compute the virtual generated columns by its dependent columns currently, and the whole virtual generated columns will be pushed down to TiKV in future.

Reply:
1. Not all the expressions can be pushed down to coprocessor, it would be more complicate if it contains virtual columns.
2. If `v_col` is defined as `a+b`, we can't rewrite it directly. For example, the type of `v_col` is `tinyint`.

Reply:
1. filters in `tableConds` are expressions that can be pushed down to coprocessor;
2. this is exactly what we do in `projectVirtualColumns` now, which considers the type casting;

IMHO this part is not that graceful because we extracted filters that can be pushed down to coprocessor, and then this filter turned out to be one that should be kept in TiDB. Also, it may incur possible performance regression for specific scenarios when these filters are not pushed down.

If you really insist current implementation, I prefer to at least leave a `TODO` here as a reminder that we should optimize this part later. Also,

> and the whole virtual generated columns will be pushed down to TiKV in future.

When would this be supported approximately?

Reply:
1. How can you make sure the expression of virtual column can be pushed down to coprocessor? The virtual columns's expression maybe contain some expressions didn't implement in cop. 
2. `projectVirtualColumns` can't deal with type tinyint, and it's one of reasons to refactoring.

I'm agree with you that this implement isn't graceful, but currently we can't compute the value of virtual column in TiKV, it seems like a workaround.

> Also, it may incur possible performance regression for specific scenarios when these filters are not pushed down.

It shouldn't be a regression, because we didn't support it before this pr.

> leave a TODO here as a reminder that we should optimize this part later

In fact，most of code in this pr will be removed and refactoring when TiKV support projection and can compute the value of virtual column. From this perspective，most of code in this pr should have a TODO, but is it seems tedious?

> When would this be supported approximately?

I think it would be supported in half a year.